### PR TITLE
BDDL refactor: unified KnowledgeBase API, explicit predicates, and clean task interface

### PR DESCRIPTION
**Pull request by cgokmen**
## Summary

Complete refactor of the BDDL library to provide a clean, unified API through a single `KnowledgeBase` entry point.

- **Domain rename**: BDDL domains renamed from `omnigibson`/`igibson` to `behavior-1k`/`behavior-100`
- **Backend removal**: Delete `BDDLBackend` system; all predicate evaluation/sampling uses user-supplied callbacks
- **Explicit predicates** (`bddl/predicates.py`): 26 concrete predicate classes (`OnTop`, `Cooked`, etc.) that are the leaf nodes of the expression tree — callbacks receive the class, not a string
- **Typed transition rules** (`bddl/transition_rules.py`): `CookingRecipe`, `MixingRecipe`, `MachineRecipe`, `SubstanceCookingRecipe`, `WasherRule` dataclasses with predicate classes resolved at parse time
- **Unified KnowledgeBase**: Single `KnowledgeBase()` entry point replaces `ObjectTaxonomy`, `bddl.task.Task`, and separate transition rule loading
- **CompiledTask**: `Task.compile(scene_layout)` returns an immutable `CompiledTask` — multiple compilations coexist for different environments
- **Wildcard expansion** (`bddl/wildcard.py`): `*` instances expanded natively in BDDL given a `{room_type: {category: count}}` scene layout; `get_processed_bddl` deleted from OmniGibson
- **Scope is a set**: Object scope is `set[str]` instead of a `dict[str, str]` identity mapping
- **BDDLEntity removed**: Object scope maps instance names directly to raw sim objects; predicate class infrastructure simplified to a `PREDICATE_TO_STATE` dict
- **KB load optimization**: Shadow set dedup + optional WordNet loading; ~5.5s to ~2.1s
- **KB dataclass refactor** (merged): Peewee ORM replaced with plain dataclasses; `PredicateUsage` renamed from `Predicate`
- **OmniGibson migration**: Single shared `KB` in `bddl_utils.py`; all ObjectTaxonomy/bddl.task/bddl.activity imports removed
- **Asset pipeline + JoyLo**: All 13 asset pipeline scripts and JoyLo utils migrated to KB API
- **Test suite + CI**: 44 pytest tests covering KB population, condition evaluation, task compilation, wildcards, concurrent compilations; added to GitHub Actions workflow

### CODE REVIEW

File: asset_pipeline/b1k_pipeline/tools/category_qa_josiah.py [hunk 1]
```diff
@@ -477,7 +477,7 @@ def evaluate_single_object(self, all_objects, i):
         print(f"\n\n\n\nNow editing object {obj.name.replace('obj_', '')}\n")
 
         # If we're checking joints and the object has none, skip
-        synset = self.taxonomy.get_synset_from_category(category=obj.category)
+        synset = self.kb.get_category(obj.category).synset.name if self.kb.get_category(obj.category) else None
```
Comment (stefren) [inline asset_pipeline/b1k_pipeline/tools/category_qa_josiah.py:479]: is `self.kb.get_category(obj.category)` ever defined but `.synset` is `None`?

File: bddl3/bddl/knowledge_base/processing.py [hunk 2]
```diff
@@ -21,519 +20,654 @@
...
+
+    debug_print(verbose, "Finished prep work...")
+    return object_taxonomy, object_rename_mapping
+
+
+def post_complete_operation():
+    """
+    put any post completion work (e.g. update stuff) here
+    """
+    # debug_print(True, "Running post completion operations...")
```
Comment (stefren) [inline bddl3/bddl/knowledge_base/processing.py]: add pass here?

File: bddl3/bddl/knowledge_base/processing.py [hunk 3]
```diff
@@ -21,519 +20,654 @@
 logger = logging.getLogger(__name__)
 
 
-class KnowledgeBaseProcessor:
-    def __init__(self, verbose=True) -> None:
-        self._verbose = verbose
+def _get_shadow_set(obj, field_name):
+    """Get or create a shadow set for O(1) membership checks during population."""
+    if not hasattr(obj, "_shadow_sets"):
```
Comment (stefren) [inline bddl3/bddl/knowledge_base/processing.py]: nit, do we know the type of these objects? would prefer to instantiate with this field instead of adding it afterwards

File: bddl3/bddl/knowledge_base/processing.py [hunk 4]
```diff
@@ -21,519 +20,654 @@
...
+    return add_fn(key), True
+
+
+def tqdm_iter(iterable, verbose, *args, **kwargs):
+    if verbose:
+        return tqdm.tqdm(iterable, *args, **kwargs)
+    else:
+        return iterable
+
+def debug_print(verbose, *args, **kwargs):
```
Comment (stefren) [inline bddl3/bddl/knowledge_base/processing.py]: shouldn't we be able to use `logger` for this?

File: bddl3/bddl/knowledge_base/processing.py [hunk 5]
```diff
@@ -21,519 +20,654 @@
...
+                try:
+                    room = kb.add_room(
+                        name=room_name,
+                        type=room_name.rsplit("_", 1)[0],
+                        scene=scene,
+                    )
+                    link_many_to_one(room, "scene", scene, "rooms")
+                except ValueError:
+                    raise Exception(
+                        f"room {room_name} in {scene.name} (not ready) already exists!"
```
Comment (stefren) [inline bddl3/bddl/knowledge_base/processing.py]: do we know this is the only reason a `ValueError` would be thrown?

File: bddl3/bddl/knowledge_base/processing.py [hunk 6]
```diff
@@ -21,519 +20,654 @@
...
+                    )
+                for orig_name, count in planned_scene_dict[scene_name][
+                    room_name
+                ].items():
+                    if orig_name.split("-")[1] not in deletion_queue:
+                        object_name = orig_name
+                        orig_id = orig_name.split("-")[1]
```
Comment (stefren) [inline bddl3/bddl/knowledge_base/processing.py]: nit: can pull this above the if statement above

File: OmniGibson/omnigibson/utils/bddl_utils.py [hunk 7]
```diff
@@ -124,85 +123,34 @@
...
-        state_class.__name__ + "StateBinaryPredicate",
-        (ObjectStateBinaryPredicate,),
-        {"STATE_CLASS": state_class, "STATE_NAME": state_name},
-    )
+def og_categories_from_bddl_inst(bddl_inst):
+    """Get OG categories for a BDDL instance."""
+    synset_obj = KB.get_synset(synset_from_bddl_inst(bddl_inst))
+    if synset_obj.is_substance:
+        return [ps.name for s in [synset_obj] + synset_obj.descendants for ps in s.particle_systems]
+    return [c.name for s in [synset_obj] + synset_obj.descendants if s.is_leaf for c in s.categories]
```
Comment (stefren) [inline OmniGibson/omnigibson/utils/bddl_utils.py]: nit: i would just pull out `[synset_obj] + synset_obj.descendants` to make this more readable

File: OmniGibson/omnigibson/learning/eval.py [hunk 8]
```diff
@@ -283,8 +283,10 @@ def load_task_instance(self, instance_id: int, test_hidden: bool = False) -> Non
         # causes some jitter (maybe for small mass / thin objects?)
         for _ in range(25):
             og.sim.step_physics()
-            for entity in self.env.task.object_scope.values():
-                if not entity.is_system and entity.exists:
+            for inst, entity in self.env.task.object_scope.items():
+                from omnigibson.utils.bddl_utils import is_system_bddl_inst
```
Comment (stefren) [inline OmniGibson/omnigibson/learning/eval.py]: nit, we should probably just have this import at the top-level? Is it especially heavy / do we expect it to not be used typically?

File: OmniGibson/omnigibson/examples/objects/view_cloth_configurations.py [hunk 9]
```diff
@@ -21,12 +21,12 @@ def main(random_selection=False, headless=False, short_exec=False):
...
     # Select a category to load
     available_obj_categories = get_all_object_categories()
-    object_taxonomy = ObjectTaxonomy()
     cloth_obj_categories = [
         category
         for category in available_obj_categories
-        if object_taxonomy.get_synset_from_category(category)
-        and "cloth" in object_taxonomy.get_abilities(object_taxonomy.get_synset_from_category(category))
+        if KB.get_category(category) is not None
+        and KB.get_category(category).synset is not None
```
Comment (stefren) [inline OmniGibson/omnigibson/examples/objects/view_cloth_configurations.py]: i feel like we use this pattern a lot, should we create a helper on the knowledge base that returns get_category_abilities or `[]` or `None` or other sane thing? or otherwise check if x in  abilities?

File: joylo/gello/utils/og_teleop_utils.py [hunk 10]
```diff
@@ -18,20 +18,19 @@
 from omnigibson.utils import transform_utils as T
 from omnigibson.sensors import VisionSensor
 from omnigibson.objects.usd_object import USDObject
-from bddl.activity import Conditions
+from gello.robots.sim_robot.og_teleop_cfg import *
```
Comment (stefren) [inline joylo/gello/utils/og_teleop_utils.py:21]: why import star? ideally just import the methods you need

File: bddl3/bddl/knowledge_base/models.py [hunk 11]
```diff
@@ -4,21 +4,23 @@
 import json
```
Comment (stefren) [inline bddl3/bddl/knowledge_base/models.py:4]: i'm not really that familiar with how we're using these models, is there a database instance somewhere? do we need to migrate data?

File: asset_pipeline/b1k_pipeline/category_qa_viewer.py [hunk 12]
```diff
@@ -26,7 +26,7 @@
     get_object_models_of_category,
 )
 
-from bddl.object_taxonomy import ObjectTaxonomy
+from bddl.knowledge_base import KnowledgeBase
```
Comment (stefren) [inline asset_pipeline/b1k_pipeline/category_qa_viewer.py]: what is this being used for?

File: asset_pipeline/b1k_pipeline/qa_viewer_og.py [hunk 13]
```diff
@@ -13,7 +13,7 @@
 nltk.download("omw-1.4")
 from nltk.corpus import wordnet as wn
 
-from bddl.object_taxonomy import ObjectTaxonomy
+from bddl.knowledge_base import KnowledgeBase
```
Comment (stefren) [inline asset_pipeline/b1k_pipeline/qa_viewer_og.py]: is this used?

**Comment by stefren** [comment]
fyi tests seem to be failing due to circular import `ImportError: cannot import name 'DatasetObject' from partially initialized module 'omnigibson.objects.dataset_object' (most likely due to a circular import) (/_work/default_slurm_12/BEHAVIOR-1K/BEHAVIOR-1K/OmniGibson/omnigibson/objects/dataset_object.py)`

File: bddl3/tests/debug_simulation_example.py [hunk 14]
```diff
@@ -1,77 +1,97 @@
-from bddl.activity import *
```
Comment (wensi-ai) [inline bddl3/tests/debug_simulation_example.py:1]: is this file an actual test?

File: joylo/gello/utils/og_teleop_utils.py (hunk 10 above)
Comment (wensi-ai) [reply joylo/gello/utils/og_teleop_utils.py:21]: This is a legacy issue that's not related to this PR actually, the cfg file includes a lot of macros so I think it'll be redundant to specify each import explicitly.

File: bddl3/bddl/knowledge_base/models.py (hunk 11 above)
Comment (cgokmen) [reply bddl3/bddl/knowledge_base/models.py:4]: no, it's all in-memory. the original input JSONs map things to synset names, as we load them we build the relationships in memory.

File: asset_pipeline/b1k_pipeline/tools/category_qa_josiah.py (hunk 1 above)
Comment (cgokmen) [reply asset_pipeline/b1k_pipeline/tools/category_qa_josiah.py:479]: No, but I will add more tests for these edge cases soon just to be safe. Originally these were allowed but they are not any more, and our data tests should guarantee that.

File: asset_pipeline/b1k_pipeline/qa_viewer_og.py (hunk 13 above)
Comment (cgokmen) [reply asset_pipeline/b1k_pipeline/qa_viewer_og.py]: nope, removed

File: asset_pipeline/b1k_pipeline/category_qa_viewer.py (hunk 12 above)
Comment (cgokmen) [reply asset_pipeline/b1k_pipeline/category_qa_viewer.py]: removed

File: OmniGibson/omnigibson/learning/eval.py (hunk 8 above)
Comment (cgokmen) [reply OmniGibson/omnigibson/learning/eval.py]: must have been a circular import originally? fixed now.

File: OmniGibson/omnigibson/utils/bddl_utils.py (hunk 7 above)
Comment (cgokmen) [reply OmniGibson/omnigibson/utils/bddl_utils.py]: done

File: bddl3/bddl/knowledge_base/processing.py (hunk 2 above)
Comment (cgokmen) [reply bddl3/bddl/knowledge_base/processing.py]: done

File: bddl3/bddl/knowledge_base/processing.py (hunk 5 above)
Comment (cgokmen) [reply bddl3/bddl/knowledge_base/processing.py]: updated message

File: bddl3/bddl/knowledge_base/processing.py (hunk 6 above)
Comment (cgokmen) [reply bddl3/bddl/knowledge_base/processing.py]: done

File: joylo/gello/utils/og_teleop_utils.py (hunk 10 above)
Comment (cgokmen) [reply joylo/gello/utils/og_teleop_utils.py:21]: 100% agreed, but out of scope here.

File: OmniGibson/omnigibson/examples/objects/view_cloth_configurations.py (hunk 9 above)
Comment (cgokmen) [reply OmniGibson/omnigibson/examples/objects/view_cloth_configurations.py]: Simplified this one here, but I think it's important that the user decides if they want a category-synset-ability, or they're OK with particle system, or how they want to handle missing an assignment.

File: bddl3/bddl/knowledge_base/processing.py (hunk 4 above)
Comment (cgokmen) [reply bddl3/bddl/knowledge_base/processing.py]: Changed.

File: bddl3/bddl/knowledge_base/processing.py (hunk 3 above)
Comment (cgokmen) [reply bddl3/bddl/knowledge_base/processing.py]: I removed this whole weird mechanism now and just made the relational fields sets.

**Review by wensi-ai — APPROVED** [review]

File: bddl3/tests/debug_simulation_example.py (hunk 14 above)
Comment (cgokmen) [reply bddl3/tests/debug_simulation_example.py:1]: no, it was for my testing. removed
